File: udemy/ds-and-a/DSAndA/StacksAndQueues/StackUsingLinkedListAsUnderlyingDS.cs
using DSAndA.LinkedLists;

namespace DSAndA.StacksAndQueues
{
    public class StackUsingLinkedListAsUnderlyingDS
    {
        private DoublyLinkedNode<int> top;

        public void Push(int element)
        {
            DoublyLinkedNode<int> newNode = new DoublyLinkedNode<int>(element);
            newNode.Prev = this.top;

            if (this.top != null) this.top.Next = newNode;

            this.top = newNode;
        }

        public int Peek()
        {
            return this.top.Data;
        }

        public int Pop()
        {
            DoublyLinkedNode<int> poppedNode = this.top;

            this.top = this.top.Prev;

            return poppedNode.Data;
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/StackAndQueues/Assignment1/StackUsingLinkedListAsUnderlyingDS.js
const DoublyLinkedNode = require("../../LinkedLists/DoublyLinkedNode");

class StackUsingLinkedListAsUnderlyingDS {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    push(data) {
        let newNode = new DoublyLinkedNode(data);
        newNode.prevNode = this.tail;

        if (!this.head) {
            this.head = newNode;
        } else {
            this.tail.nextNode = newNode;
        }

        this.tail = newNode;
    }

    pop() {
        const itemToPop = this.tail;

        this.tail.prevNode.nextNode = null;
        this.tail = this.tail.prevNode;

        return itemToPop.data;
    }

    peek() {
        return this.tail.data;
    }

    toString() {
        let result = [];
        let current = this.head;

        while (current != null) {
            result.push(current.data);

            current = current.nextNode;
        }

        return result.join(', ');
    }
}

module.exports = StackUsingLinkedListAsUnderlyingDS;